clique-grpc/src/server.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{
+    proto::{self, server},
+    Error,
+};
+use clique::transport::{Request, Server};
+use futures::sync::oneshot as oneshot01;
+use futures::{Future as Future01, Stream as Stream01};
+use futures03::{future, TryFutureExt};
+use std::net::SocketAddr;
+use tokio_sync::{mpsc, oneshot};
+use tokio_tcp::TcpListener;
+use tower_grpc::{Code, Status};
+use tower_hyper::server::Http;
+
+pub struct GrpcServer {
+    msg_rx: Option<mpsc::Receiver<Result<Request, Error>>>,
+    target_tx: Option<oneshot01::Sender<SocketAddr>>,
+}
+
+#[derive(Clone)]
+pub struct Svc {
+    msg_tx: mpsc::Sender<Result<Request, Error>>,
+}
+
+pub struct Background {
+    inner: Option<Svc>,
+    target_rx: oneshot01::Receiver<SocketAddr>,
+    state: State,
+}
+
+enum State {
+    Waiting,
+    Running(Box<dyn Future01<Item = (), Error = ()> + Send>),
+}
+
+impl GrpcServer {
+    pub fn new() -> (Self, Background) {
+        let (msg_tx, msg_rx) = mpsc::channel(100);
+        let (target_tx, target_rx) = oneshot01::channel();
+
+        let server = GrpcServer {
+            msg_rx: Some(msg_rx),
+            target_tx: Some(target_tx),
+        };
+        let svc = Svc { msg_tx };
+
+        let bg = Background {
+            inner: Some(svc),
+            target_rx,
+            state: State::Waiting,
+        };
+
+        (server, bg)
+    }
+}
+
+impl Server<SocketAddr> for GrpcServer {
```

Comment:
Wouldn't `ToSocketAddr` be more flexible here?

Reply:
I went with a more specific socketaddr for now. Who ever creates this socketaddr can choose to use that or say trust-dns etc.